src/libraries/System.IO.Compression/src/System/IO/Compression/AesStream.cs

```diff
@@ -0,0 +1,316 @@
+        private bool _headerRead;
+        private long _position;
+        private readonly ReadOnlyMemory<char> _password;
+        private bool _disposed;
+
+        public AesStream(Stream baseStream, ReadOnlyMemory<char> password, bool encrypting, int keySizeBits = 256, bool ae2 = true, uint? crc32 = null)
+        {
+            ArgumentNullException.ThrowIfNull(baseStream);
+
+
+            _baseStream = baseStream;
+            _password = password;
+            _encrypting = encrypting;
+            _keySizeBits = keySizeBits;
+            _ae2 = ae2;
+            _crc32ForHeader = crc32;
+#pragma warning disable CA1416 // HMACSHA1 is available on all platforms
+            _aes = Aes.Create();
+#pragma warning restore CA1416
+            _aes.Mode = CipherMode.ECB;
+            _aes.Padding = PaddingMode.None;
+
+#pragma warning disable CA1416 // HMACSHA1 available on all platforms ?
+#pragma warning disable CA5350 // Do Not Use Weak Cryptographic Algorithms ?
+            _hmac = new HMACSHA1();
+#pragma warning restore CA5350 // Do Not Use Weak Cryptographic Algorithms
+#pragma warning restore CA1416
+
+            if (_encrypting)
+            {
+                GenerateKeys();
+                InitCipher();
+            }
+        }
+
+        private void GenerateKeys()
+        {
+            int saltSize = _keySizeBits / 16; // 8 for AES-128, 12 for AES-192, 16 for AES-256
+            _salt = new byte[saltSize];
+            RandomNumberGenerator.Fill(_salt);
+
+            // WinZip AES uses SHA1 for PBKDF2
+            byte[] derivedKey = Rfc2898DeriveBytes.Pbkdf2(_password.Span, _salt, 1000, HashAlgorithmName.SHA1, (_keySizeBits / 8) + 32 + 2);
+
+            _key = new byte[_keySizeBits / 8];
+            _hmacKey = new byte[32];
+            _passwordVerifier = new byte[2];
+
+            Buffer.BlockCopy(derivedKey, 0, _key, 0, _key.Length);
+            Buffer.BlockCopy(derivedKey, _key.Length, _hmacKey, 0, _hmacKey.Length);
+            Buffer.BlockCopy(derivedKey, _key.Length + _hmacKey.Length, _passwordVerifier, 0, _passwordVerifier.Length);
+
+            _hmac.Key = _hmacKey;
+        }
+
+        private void InitCipher()
+        {
+            if (_key is null)
+                throw new InvalidOperationException("Keys have not been generated.");
+
+            _aes.Key = _key;
+            _aesEncryptor = _aes.CreateEncryptor();
+        }
+
+        private void WriteHeader()
+        {
+            if (_headerWritten) return;
+
+            if (_salt is null || _passwordVerifier is null)
+                throw new InvalidOperationException("Keys have not been generated.");
+
+            _baseStream.Write(_salt);
+            _baseStream.Write(_passwordVerifier);
+
+            if (_ae2 && _crc32ForHeader.HasValue)
+            {
+                Span<byte> crcBytes = stackalloc byte[4];
+                BitConverter.TryWriteBytes(crcBytes, _crc32ForHeader.Value);
+                _baseStream.Write(crcBytes);
```

Comment:
Do we need to worry about endianness of the system this runs on? I assume the format uses specific endianness of the data, so using something like `BinaryPrimitives.WriteInt32BigEndian` would be more appropriate